File: sdk/typescript/src/binary.ts
import { accessSync, constants, existsSync } from "node:fs";
import { delimiter, dirname, join } from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";
import { BinaryNotFoundError } from "./errors.js";

let override: string | undefined;

/**
 * Force the SDK to use a specific `bsdkrun` binary, bypassing discovery.
 * Handy in tests or when running against a locally built debug binary.
 */
export function setBinaryPath(path: string): void {
  override = path;
  resolved = undefined;
}

/** Cross-runtime PATH lookup for an executable (Node/Deno/Bun). */
function which(name: string): string | undefined {
  const path = process.env.PATH ?? "";
  const exts = process.platform === "win32"
    ? (process.env.PATHEXT ?? ".EXE;.CMD;.BAT").split(";")
    : [""];
  for (const dir of path.split(delimiter)) {
    if (!dir) continue;
    for (const ext of exts) {
      const full = join(dir, name + ext);
      try {
        accessSync(full, constants.X_OK);
        return full;
      } catch {
        // not here / not executable — keep looking
      }
    }
  }
  return undefined;
}

/** Candidate locations, in priority order. */
function candidates(): string[] {
  const out: string[] = [];
  if (override) out.push(override);

  const env = process.env.BSDKRUN_BIN;
  if (env) out.push(env);

  // A `bsdkrun` already on PATH wins over in-repo builds.
  const onPath = which("bsdkrun");
  if (onPath) out.push(onPath);

  // In-repo dev: this file lives at sdk/typescript/src/binary.ts, so the repo
  // root is three levels up. Prefer a release build, fall back to debug.
  const here = dirname(fileURLToPath(import.meta.url));
  const repoRoot = join(here, "..", "..", "..");
  out.push(join(repoRoot, "target", "release", "bsdkrun"));
  out.push(join(repoRoot, "target", "debug", "bsdkrun"));

  return out;
}

let resolved: string | undefined;

/** Resolve (and cache) the path to the `bsdkrun` binary, or throw. */
export function resolveBinary(): string {
  if (resolved) return resolved;

  const searched = candidates();
  for (const c of searched) {
    if (c.includes("/") || c.includes("\\")) {
      if (existsSync(c)) {
        resolved = c;
        return c;
      }
    } else if (which(c)) {
      resolved = c;
      return c;
    }
  }
  throw new BinaryNotFoundError(searched);
}

/** Reset cached discovery state (mainly for tests). */
export function resetBinaryCache(): void {
  resolved = undefined;
  override = undefined;
}

File: sdk/scala/src/main/scala/bsdkrun/Binary.scala
package bsdkrun

import java.io.File
import java.nio.file.{Files, Path, Paths}

/** Locates the `bsdkrun` binary on the host and caches the result.
  *
  * Resolution order (first match wins):
  *
  *   1. an explicit override set via [[setPath]]
  *   1. the `BSDKRUN_BIN` environment variable
  *   1. `bsdkrun` on `PATH`
  *   1. an in-repo dev build: `<repo>/target/release/bsdkrun`, then
  *      `<repo>/target/debug/bsdkrun`
  *
  * [[candidates]] and [[resolveWith]] take an explicit [[Env]] so the discovery
  * logic is unit-testable without mutating real process state — the JVM has no
  * portable way to change its own environment, unlike Ruby's `ENV[...]=`.
  */
object Binary:

  /** The host state discovery reads, captured so tests can supply their own. */
  final case class Env(
      override_ : Option[String],
      bsdkrunBin: Option[String],
      path: Option[String],
      repoRoot: Option[Path]
  )

  @volatile private var overridePath: Option[String] = None
  @volatile private var resolved: Option[String] = None

  /** Force the SDK to use a specific `bsdkrun` binary, bypassing discovery. */
  def setPath(path: String): Unit = synchronized:
    overridePath = Some(path)
    resolved = None

  /** The current explicit override, if any. */
  def overridden: Option[String] = overridePath

  /** Drop cached discovery state and any override. Mainly for tests. */
  def reset(): Unit = synchronized:
    overridePath = None
    resolved = None

  private def nonEmpty(s: String | Null): Option[String] =
    Option(s).map(_.nn).filter(_.nonEmpty)

  /** The real host state, for the no-argument entry points. */
  private def hostEnv: Env = Env(
    override_ = overridePath,
    bsdkrunBin = nonEmpty(System.getenv("BSDKRUN_BIN")),
    path = nonEmpty(System.getenv("PATH")),
    repoRoot = repoRoot
  )

  /** The monorepo root, when this class is running from a source checkout
    * rather than a packaged jar: `sdk/scala/target/...` walks back up to the
    * repo. Mirrors the same dev-build fallback the other SDKs have, which is
    * what makes `sbt test` in a checkout find the freshly built CLI.
    */
  private def repoRoot: Option[Path] =
    val marker = "/sdk/scala/"
    val here = Option(getClass.getProtectionDomain.nn.getCodeSource)
      .flatMap(cs => Option(cs.nn.getLocation))
      .map(_.nn.getPath.nn)
    here.flatMap: p =>
      val idx = p.indexOf(marker)
      if idx >= 0 then Some(Paths.get(p.substring(0, idx)).nn) else None

  private def isExecutableFile(p: Path): Boolean =
    Files.isRegularFile(p) && Files.isExecutable(p)

  /** Cross-platform `PATH` lookup. `pathEnv` is a raw `PATH`-style string. */
  private def which(name: String, pathEnv: Option[String]): Option[String] =
    pathEnv.getOrElse("").split(File.pathSeparator).iterator
      .filter(_.nonEmpty)
      .map(dir => Paths.get(dir, name).nn)
      .find(isExecutableFile)
      .map(_.toString)

  /** Candidate locations, in priority order. A pure function of `env`. */
  def candidates(env: Env): Seq[String] =
    val builder = Seq.newBuilder[String]
    env.override_.foreach(builder += _)
    env.bsdkrunBin.foreach(builder += _)
    which("bsdkrun", env.path).foreach(builder += _)
    env.repoRoot.foreach: root =>
      builder += root.resolve("target/release/bsdkrun").nn.toString
      builder += root.resolve("target/debug/bsdkrun").nn.toString
    builder.result()

  /** Candidate locations for the real host. */
  def candidates(): Seq[String] = candidates(hostEnv)

  /** Resolve the binary against an explicit environment, without caching. */
  def resolveWith(env: Env): Either[BsdkrunError, String] =
    val searched = candidates(env)
    val found = searched.find: candidate =>
      if candidate.contains(File.separator) then Files.exists(Paths.get(candidate).nn)
      else which(candidate, env.path).isDefined
    found.toRight(BsdkrunError.BinaryNotFound(searched))

  /** Resolve (and cache) the path to the `bsdkrun` binary. */
  def resolve(): Either[BsdkrunError, String] =
    resolved match
      case Some(path) => Right(path)
      case None =>
        val out = resolveWith(hostEnv)
        out.foreach(p => synchronized { resolved = Some(p) })
        out
